### PR TITLE
Retry token fixes

### PR DESCRIPTION
**Pull request by Ralith**
Resolves some correctness issues detected in interop testing, and makes us a bit better at detecting them ourselves.

### CODE REVIEW

File: quinn-proto/src/connection/mod.rs [hunk 1]
```diff
@@ -1908,6 +1917,23 @@ impl Connection {
                     trace!("dropping short packet during handshake");
                     return;
                 } else {
+                    if let (
+                        &Header::Initial { ref token, .. },
+                        &State::Handshake(state::Handshake {
+                            token: ref expected_token,
+                            ..
+                        }),
+                    ) = (&packet.header, &self.state)
```
Comment (djc) [inline quinn-proto/src/connection/mod.rs:1934]: Oof, I often find multi-line `if let`s prety hard to follow. Is there some other way we can format this, maybe using some temporary named variables?

**Review by djc — COMMENTED** [review]
I'm confused by all this. This seems to replace `Option`s with empty byte sequences all over the place, which doesn't seem like an improvement, especially when we just add comments saying they're always empty for clients.

File: quinn-proto/src/connection/mod.rs [hunk 2]
```diff
@@ -243,7 +243,7 @@ impl Connection {
         };
         let state = State::Handshake(state::Handshake {
             rem_cid_set: side.is_server(),
-            token: None,
+            token: Bytes::new(),
```
[the remark this answers is not in the document]
Comment (Ralith) [reply quinn-proto/src/connection/mod.rs]: Because `None` is always handled exactly as if it were the empty `Bytes`. It's not a sentinel value, because we don't detect it and thereby special-case behavior: because the normal case is literally an empty token, the empty `Bytes` produces the correct behavior without special effort.

**Review by djc — DISMISSED** [review]

File: quinn-proto/src/connection/mod.rs (hunk 1 above)
Comment (Ralith) [reply quinn-proto/src/connection/mod.rs]: Split into two `if let`s and dropped the `state::Handshake` destructuring. Much clearer, I agree.

**Review by djc — APPROVED** [review]
